Graphemic lexicon updates / fixes in babel/s5d recipe and hub4_spanish recipe

m-wiesner opened the pull request:
This updates the graphemic lexicon scripts so that they are a little easier to use. Changes include:

- minor differences in handling of special characters
- separate scripts for morfessor training and segmentation in babel s5d recipe
- added option in make_unicode_lexicon.py [--apply-map] for adding lexicon entries for new words
- udpates graphemic lexicon documentation
- adds egs/hub4_spanish/local/expand_graphemic_lexicon.sh which creates the dict and lang 
  directories from a new list of words
- fixes egs/hub4_spanish/local/prepare_lexicon.sh to use updated graphemic lexicon scripts.

danpovey commented:
@jtrmal, I'll wait for you to sign off on this before checking it further.
I'm not sure that I like the all-caps filenames.  is that really the first,
most important that we want people entering the babel/s5c directory to see?

On Wed, Jul 5, 2017 at 5:43 PM, m-wiesner <notifications@github.com> wrote:

> This updates the graphemic lexicon scripts so that they are a little
> easier to use. Changes include:
>
>    - minor differences in handling of special characters
>    - separate scripts for morfessor training and segmentation in babel
>    s5d recipe
>    - added option in make_unicode_lexicon.py [--apply-map] for adding
>    lexicon entries for new words
>    - udpates graphemic lexicon documentation
>    - adds egs/hub4_spanish/local/expand_graphemic_lexicon.sh which
>    creates the dict and lang
>    directories from a new list of words
>    - fixes egs/hub4_spanish/local/prepare_lexicon.sh to use updated
>    graphemic lexicon scripts.
>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/kaldi-asr/kaldi/pull/1740
> Commit Summary
>
>    - Updates babel graphemic lexicon recipe
>    - Adds example script for generating gaphemic lexicons.
>    - Adds train and apply scripts for morfessor word segmentation
>    - Adds local/lexicon/make_word_list.py
>    - Minor fix to run-1-main-unicode-extend-lex.sh
>    - Updates hub4_spanish prepare_lexicon.sh to match updated graphemic
>    lexicon scripts.
>    - Updates hub4_spanish prepare_lexicon.sh to match updated graphemic
>    lexicon scripts. Adds expand_graphemic_lexicon.sh to add new words to a
>    lexicon
>
> File Changes
>
>    - *M* egs/babel/s5d/RUN_UNICODE_SYSTEM
>    <https://github.com/kaldi-asr/kaldi/pull/1740/files#diff-0> (2)
>    - *M* egs/babel/s5d/UNICODE_README
>    <https://github.com/kaldi-asr/kaldi/pull/1740/files#diff-1> (109)
>    - *A* egs/babel/s5d/local/lexicon/apply_morphs.sh
>    <https://github.com/kaldi-asr/kaldi/pull/1740/files#diff-2> (32)
>    - *M* egs/babel/s5d/local/lexicon/make_unicode_lexicon.py
>    <https://github.com/kaldi-asr/kaldi/pull/1740/files#diff-3> (344)
>    - *M* egs/babel/s5d/local/lexicon/make_word_list.py
>    <https://github.com/kaldi-asr/kaldi/pull/1740/files#diff-4> (9)
>    - *A* egs/babel/s5d/local/lexicon/train_morphs.sh
>    <https://github.com/kaldi-asr/kaldi/pull/1740/files#diff-5> (27)
>    - *M* egs/babel/s5d/local/prepare_unicode_lexicon.py
>    <https://github.com/kaldi-asr/kaldi/pull/1740/files#diff-6> (276)
>    - *M* egs/babel/s5d/run-1-main-unicode-extend-lex.sh
>    <https://github.com/kaldi-asr/kaldi/pull/1740/files#diff-7> (44)
>    - *M* egs/babel/s5d/run-1-main-unicode.sh
>    <https://github.com/kaldi-asr/kaldi/pull/1740/files#diff-8> (46)
>    - *A* egs/babel/s5d/unicode_lexicon_examples.sh
>    <https://github.com/kaldi-asr/kaldi/pull/1740/files#diff-9> (112)
>    - *A* egs/hub4_spanish/s5/local/expand_graphemic_lexicon.sh
>    <https://github.com/kaldi-asr/kaldi/pull/1740/files#diff-10> (64)
>    - *M* egs/hub4_spanish/s5/local/lexicon/make_unicode_lexicon.py
>    <https://github.com/kaldi-asr/kaldi/pull/1740/files#diff-11> (308)
>    - *M* egs/hub4_spanish/s5/local/prepare_lexicon.sh
>    <https://github.com/kaldi-asr/kaldi/pull/1740/files#diff-12> (26)
>    - *M* egs/hub4_spanish/s5/local/prepare_unicode_dict.py
>    <https://github.com/kaldi-asr/kaldi/pull/1740/files#diff-13> (274)
>
> Patch Links:
>
>    - https://github.com/kaldi-asr/kaldi/pull/1740.patch
>    - https://github.com/kaldi-asr/kaldi/pull/1740.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/1740>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVu1dIMp5Nz9BK647sJ0dmWp_4Oon-ks5sLAN9gaJpZM4OO6Cf>
> .
>

m-wiesner commented:
I agree, but I figured I'd just kept the file names that were already there.

danpovey commented:
oh I didn't notice they were preexisting.

On Wed, Jul 5, 2017 at 8:11 PM, m-wiesner <notifications@github.com> wrote:

> I agree, but I figured I'd just kept the file names that were already
> there.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/1740#issuecomment-313260243>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVu_UnlWhYBlPQPqpiWhp3ATikZnOAks5sLCZCgaJpZM4OO6Cf>
> .
>

m-wiesner commented:
We probably should change it though. We'll see what Yenda has to say.

jtrmal commented on egs/babel/s5d/local/lexicon/apply_morphs.sh:
```diff
@@ -0,0 +1,32 @@
+  echo "Usage: ./local/apply_morphs.sh [opts] <morphs_model_dir> <wordlist> <_word2baseform>"
+  echo "  --encoding <encoding_type> # lexicon encoding"
+  echo "  --max-len <int> # maximum allowed morpheme length"  
+  exit 1
```
generally speaking, when printing error mesages, the echos should be 
```
echo >&2 "something"
```
typically, we are not extremely careful about it but I think we perhaps should be.

jtrmal commented on egs/babel/s5d/local/lexicon/train_morphs.sh:
```diff
@@ -0,0 +1,27 @@
+if [ $# -ne 2 ]; then
+  echo "Usage: ./local/train_morphs.sh [opts] <path_to_wordcounts> <morphs_directory>" 
+  echo "  --encoding <encoding_type> # lexicon encoding" 
+  exit 1
```
the same about echos

jtrmal commented on egs/babel/s5d/run-1-main-unicode.sh:
```diff
@@ -96,6 +97,7 @@ if [[ ! -f data/lang/L.fst || data/lang/L.fst -ot $lexicon ]]; then
     data/local $oovSymbol data/local/tmp.lang data/lang
 fi
 
+exit
```
this is probably mistake?

m-wiesner replied:
Yeah. Good catch. I was testing the lexicon portion so I added an exit that I forgot to get rid of.

jtrmal commented on egs/hub4_spanish/s5/local/expand_graphemic_lexicon.sh:
```diff
@@ -0,0 +1,64 @@
+
+. ./utils/parse_options.sh
+
+if [ $# -eq 0 ]; then
```
`$# -ne 6`

jtrmal commented on egs/hub4_spanish/s5/local/prepare_unicode_dict.py:
```diff
@@ -3,11 +3,19 @@
+# ======= Prepare dictionary directory (e.g. data/local) from lexicon.txt =====
+# This script creates all files in the dictionary directory.
 #
+# This script basically takes the place of the prepare_lexicon.pl script for
```
I deleted this intentionally because no-one cares how is it done in babel.
I'd say put the original text back and tweak it to taste (list the files to be created if you want). Keep the reference to lexicon/make_unicode_lexicon.py

jtrmal commented on egs/babel/s5d/local/prepare_unicode_lexicon.py:
```diff
@@ -3,20 +3,19 @@
+# 3. optional_silence.txt
+# 4. extra_questions.txt
+#
+# It requires only a valid kaldi format lexicon (lexicon.txt) as input
```
add the reference to local/lexicon/make_unicode_lexicon.py

jtrmal commented on egs/babel/s5d/local/prepare_unicode_lexicon.py:
```diff
@@ -25,173 +24,170 @@
 
-# Extract a sorted set of distinct unicode graphemes from the lexicon
-def extract_graphemes(table):
+# Extract a sorted set of distinct phonemes from the lexicon
```
when I'm looking at the changes, perhaps important part for a person reading those script would be to understand the distinction between graphemes and phonemes -- can you please document that? Another thing to consider if we want the acoustic units phonemes (but I think that would be ok, if it's documented what that means).

m-wiesner commented:
Alright. I believe I've committed all the changes you mentioned.

jtrmal commented:
Switch to "Files changed" and try to review all my comments that are still there, please.

jtrmal commented:
Scratch that -- as far as I can see, you indeed addressed everything. Sorry. I'll go through it one more time.

jtrmal commented on egs/babel/s5d/local/lexicon/make_word_list.py:
```diff
@@ -59,8 +58,8 @@ def process_transcripts(transcripts_dir, transcripts_list):
 
 def main():
     if len(sys.argv[1:]) == 0:
```
I wonder if you could write len(sys.argv) == 1:

jtrmal commented on egs/babel/s5d/unicode_lexicon_examples.sh:
```diff
@@ -0,0 +1,112 @@
+#!/bin/bash
```
I'm not sure about the purpose of this file? Could you document it's purpose?

m-wiesner replied:
It was meant as part of the readme_unicode file, and it seems I forgot to merge those two files. Is it ok if I put these examples in that file? I wanted to put an example of how to add entries to the lexicon from new external text that you get at a later date without having to retrain the acoustic models, but otherwise, most of these examples are in run-1-main-unicode.sh

jtrmal commented on egs/hub4_spanish/s5/local/expand_graphemic_lexicon.sh:
```diff
@@ -0,0 +1,64 @@
+#! /bin/bash
```
is this file really needed in the recipe? Or is it just something you wrote for Chunxi?

m-wiesner replied:
I guess not. It's just there to properly generate a new lexicon from extra text when you have already trained the acoustic models, which was what Chunxi was doing. Should I get rid of it?

jtrmal commented:
I'd suggest to add some heavily simplified example you feel it might be
useful into the readme but remove this file. Those examples can be pseudo
code, just to show the process, not line-by-line copy and paste code..
Y.

On Jul 17, 2017 6:12 PM, "m-wiesner" <notifications@github.com> wrote:

> *@m-wiesner* commented on this pull request.
> ------------------------------
>
> In egs/babel/s5d/unicode_lexicon_examples.sh
> <https://github.com/kaldi-asr/kaldi/pull/1740#discussion_r127742983>:
>
> > @@ -0,0 +1,112 @@
> +#!/bin/bash
>
> It was meant as part of the readme_unicode file, and it seems I forgot to
> merge those two files. Is it ok if I put these examples in that file? I
> wanted to put an example of how to add entries to the lexicon from new
> external text that you get at a later date without having to retrain the
> acoustic models, but otherwise, most of these examples are in
> run-1-main-unicode.sh
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/1740#discussion_r127742983>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AKisXxAk_iawe55A_SRV6RY8yA9jnlZCks5sO4Z4gaJpZM4OO6Cf>
> .
>

jtrmal commented:
Yes, please remove it. Y.

On Jul 17, 2017 6:09 PM, "m-wiesner" <notifications@github.com> wrote:

> *@m-wiesner* commented on this pull request.
> ------------------------------
>
> In egs/hub4_spanish/s5/local/expand_graphemic_lexicon.sh
> <https://github.com/kaldi-asr/kaldi/pull/1740#discussion_r127741756>:
>
> > @@ -0,0 +1,64 @@
> +#! /bin/bash
>
> I guess not. It's just there to properly generate a new lexicon from extra
> text when you have already trained the acoustic models, which was what
> Chunxi was doing. Should I get rid of it?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/1740#discussion_r127741756>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AKisXwlbDRMsGVYWyYn8GGSDLAkkjB3jks5sO4XIgaJpZM4OO6Cf>
> .
>

m-wiesner commented:
I removed both files, and just modified readme_unicode a little to incorporate the one part of the unicode_lexicon_examples.sh that I wanted to mention.

jtrmal commented:
OK, did you run the hub4_spanish just to see it trains and the results are reasonable? If not, can you please do it?

jtrmal commented:
in any case, update the results file

m-wiesner commented:
Ok. tri5 is finished. I just ran decode_fmllr_extra.sh and I'm reporting whatever scoring that did.

Previous: 22.92%
My run: 22.53%    When I look at the same LM weight as the previous number it's 22.74%

So it seems to checks out. I'm running the chain tdnn. When you say update the results file, did you mean all the numbers reported in there including the sgmm systems?

jtrmal commented:
Yes, please update (overwrite) all the results
Y.

On Jul 18, 2017 8:16 PM, "m-wiesner" <notifications@github.com> wrote:

> Ok. tri5 is finished. I just ran decode_fmllr_extra.sh and I'm reporting
> whatever scoring that did.
>
> Previous: 22.92%
> My run: 22.53% When I look at the same LM weight as the previous number
> it's 22.74%
>
> So it seems to checks out. I'm running the chain tdnn. When you say update
> the results file, did you mean all the numbers reported in there including
> the sgmm systems?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/1740#issuecomment-316150490>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AKisX81w7JgYj69szCIJoAzz0tLphsrrks5sPPZ9gaJpZM4OO6Cf>
> .
>

jtrmal commented:
@danpovey  this is ready for merge, I believe
